cardano-crypto-class/src/Cardano/Crypto/Hash/Packed.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE DeriveGeneric #-}
+module Cardano.Crypto.Hash.Packed where
+
+import qualified Data.Primitive.ByteArray as BA
+import qualified Data.ByteString.Short as SBS
+import           Data.ByteString.Short (ShortByteString)
+import qualified Data.ByteString.Short.Internal as SBSI
+import           NoThunks.Class (NoThunks)
+import           GHC.Generics (Generic)
+import           Data.Word (Word64, Word8)
+import qualified Control.Monad.ST as ST
+import           Control.DeepSeq (NFData)
+
+data Bytes32 = Bytes32
+    {-# UNPACK #-} !Word8 -- length
```

Comment:
This will add not 1 byte extra. but 8 bytes. because `data Word8 = W8# Word#` and because it will be aligned to the bigger `Word64` type that follows anyways.

Reply:
True. I'm not sure why you're telling me this.

Reply:
I am telling you this because that is the overhead that we can avoid,

Reply:
I mean that I thought we already both know the amount of the overhead this approach would involve.

I'm not sure zero overhead is possible without a drastic impact on upstream uses. Your alternative has the same drastic impact as the original, and for the same reasons.

Reply:
Define "drastic impact on downstream uses". I don't think going through the compiler errors and adding a constraint is such a big of a deal. Especially since we want to squeeze as much savings from this as possible, don't we?